### PR TITLE
Add error handling to iterators

### PR DESCRIPTION
**Pull request by adaszko**
Before this, iterators would simply stop (i.e. produce `None`) on an IO error for instance.  With this PR, a proper error is produced.

### CODE REVIEW

File: beacon_node/store/src/forwards_iter.rs [hunk 1]
```diff
@@ -129,33 +131,36 @@ impl<E: EthSpec> Iterator for HybridForwardsBlockRootsIterator<E> {
...
-                            continuation_data.take().or_else(|| {
-                                error!(
-                                    iter.inner.store.log,
-                                    "HybridForwardsBlockRootsIterator: logic error"
-                                );
-                                None
-                            })?;
+                        let (end_state, end_block_root) = continuation_data
+                            .take()
+                            .expect("HybridForwardsBlockRootsIterator: logic error");
```
[the remark this answers is not in the document]
Comment (michaelsproul) [reply beacon_node/store/src/forwards_iter.rs:141]: Hmm, GitHub didn't combine our comments, but I think this should return an `Err` seeing as we're already under a `Result`

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 2]
```diff
@@ -635,8 +657,12 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
     pub fn root_at_slot(&self, target_slot: Slot) -> Result<Option<Hash256>, Error> {
         Ok(self
             .rev_iter_block_roots()?
-            .find(|(_root, slot)| *slot == target_slot)
-            .map(|(root, _slot)| root))
+            .find(|result| match result {
+                Ok((_, slot)) => *slot == target_slot,
+                Err(_) => true,
+            })
```
[the remark this answers is not in the document]
Comment (adaszko) [reply beacon_node/beacon_chain/src/beacon_chain.rs:663]: I think hiding `Err(_) => true` behind a function call is going to lead to problems.  `true` within `find()` signifies the value has been found and `Err(_)` almost never is the searched value.

Attempting abstracting from a slightly different angle can produce slightly cleaner code.  That angle would be [`itertools::process_results`](https://docs.rs/itertools/0.9.0/itertools/fn.process_results.html).  For instance, translating the piece of code above:

```Rust
use itertools::process_results;

pub fn root_at_slot(&self, target_slot: Slot) -> Result<Option<Hash256>, Error> {
    process_results(self.rev_iter_state_roots()?, |mut iter| {
        iter.find(|(_, slot)| *slot == target_slot).map(|(root, _)| root)
    })
}
```

It has the advantage that you can write your iterators pipeline as if there were no `Err`s in it *and* you get rid of the `.transpose()` call.  The assumptions that's made is that the pipeline gets 'folded' into a single value, and not into another iterator.

Another example with `take_while()` for comparison:

Before:

```Rust
let state_root = self
    .rev_iter_state_roots()?
    .take_while(|result| match result {
        Ok((_, current_slot)) => *current_slot >= slot,
        Err(_) => true,
    })
    .find(|result| match result {
        Ok((_, current_slot)) => *current_slot == slot,
        Err(_) => true,
    })
    .transpose()?
    .map(|(root, _slot)| root)
    .ok_or_else(|| Error::NoStateForSlot(slot))?;
```

After:

```Rust
let state_root = process_results(self.rev_iter_state_roots()?, |iter| {
    iter.take_while(|(_, current_slot)| *current_slot >= slot)
        .find(|(_, current_slot)| *current_slot == slot)
        .map(|(root, _slot)| root)
})?.ok_or_else(|| Error::NoStateForSlot(slot))?;
```

**Comment by adaszko** [comment]
> Regarding Paul's point about infinite iteration, I think we discussed this previously and decided that collect will short-circuit on the first Err?

That's right! @paulhauner https://github.com/sigp/lighthouse/pull/1132#discussion_r424288514

**Comment by adaszko** [comment]
FYI, [those currently failing tests](https://github.com/adaszko/lighthouse/runs/742406667) pass for me on macOS.  I will have to run them on my Linux machine.

**Comment by paulhauner** [comment]
> Regarding Paul's point about infinite iteration, I think we discussed this previously and decided that `collect` will short-circuit on the first `Err`?

It depends on the type that you `collect` into. See this playground for an example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=27415619b28bdad42c3f70f573561da5

File: beacon_node/store/src/forwards_iter.rs [hunk 3]
```diff
@@ -63,22 +63,26 @@ impl SimpleForwardsBlockRootsIterator {
...
-        }
+        let values = std::iter::once(Ok((end_block_root, end_state.slot)))
+            .chain(BlockRootsIterator::owned(store, end_state))
+            .take_while(|result| match result {
+                Ok((_, slot)) => *slot >= start_slot,
+                Err(_) => true,
+            })
+            .collect::<Result<Vec<_>>>()?;
+        let result = Self { values: values };
+        Ok(result)
```
[the remark this answers is not in the document]
Comment (michaelsproul) [reply beacon_node/store/src/forwards_iter.rs]: That's true, I normally refactor during debugging, or use `dbg!`

**Pull request merged**